### PR TITLE
CFE-2546: Extended hostsseen() policy function to return host keys

### PR DESCRIPTION
**Pull request by KotzaBoss**
Ticket: CFE-2546
Changelog: Title

### CODE REVIEW

**Comment by KotzaBoss** [comment]
How can [this](https://github.com/cfengine/core/blob/master/libpromises/evalfunction.c#L522-L526) be done by avoiding the `sprintf`/`sscanf` combo that starts [here](https://github.com/cfengine/core/blob/master/libpromises/evalfunction.c#L630)?

How would it look like if lastseen is stored directly as a `time_t`?

File: libpromises/evalfunction.c [hunk 1]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+    for (const Rlist *arg = finalargs; arg; arg = arg->next)
+    {
+        if (!IsDefinedClass(ctx, RlistScalarValue(arg)))
+        {
+            return FnReturnContext(false);
+        }
+    }
+
+    return FnReturnContext(true);
+}
```
Comment (olehermanse) [inline libpromises/evalfunction.c:538]: Why did you move FnCallAnd? If there is a good reason, do it in a separate commit and explain why.

File: libpromises/evalfunction.c [hunk 2]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+typedef enum {
+    NAME,
+    ADDRESS,
+    KEY,
+    NONE
+} HostsSeenFormatOption;
+
+static HostsSeenFormatOption ValidateHostsSeenFormatOption(const char *format)
+{
+    if (strcmp(format, "name") == 0)
```
Comment (olehermanse) [inline libpromises/evalfunction.c:579]: `StringEqual()`

File: libpromises/evalfunction.c [hunk 3]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+
+/************** FnCallHostsSeen utilities **************/
+
+typedef struct {
+    char *address;
+    char *hostkey;
+    char *lastseen;
+} HostData;
+
+static HostData *HostDataNew(const char *address, const char *hostkey, const char *buf)
```
Comment (olehermanse) [inline libpromises/evalfunction.c:548]: Why call it buf if it's lastseen?

File: libpromises/evalfunction.c [hunk 4]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+
+    return FnReturnContext(true);
+}
+
+/************** FnCallHostsSeen utilities **************/
+
+typedef struct {
+    char *address;
+    char *hostkey;
+    char *lastseen;
```
Comment (olehermanse) [inline libpromises/evalfunction.c:545]: Why not use the proper type?

File: libpromises/evalfunction.c [hunk 5]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+    data->lastseen = xstrdup(buf);
+    return data;
+}
+
+static void HostDataFree(HostData *hd)
+{
+    assert(hd != NULL);
+    assert(hd->address != NULL
+           && hd->hostkey != NULL
+           && hd->lastseen != NULL);
```
Comment (olehermanse) [inline libpromises/evalfunction.c:562]: Our normal approach to destroy functions is to mimic free. `free(NULL)` is a no-op. Any reason to do something different?

```suggestion
    if (hd == NULL)
    {
        return;
    }
```

Comment (olehermanse) [reply libpromises/evalfunction.c:562]: https://github.com/cfengine/core/blob/master/CONTRIBUTING.md#safety--correctness

> `Destroy` functions should accept `NULL` pointers (similar to `free()`).

File: libpromises/evalfunction.c [hunk 6]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+    NONE
+} HostsSeenFormatOption;
+
+static HostsSeenFormatOption ValidateHostsSeenFormatOption(const char *format)
+{
+    if (strcmp(format, "name") == 0)
+    {
+        return NAME;
+    }
+    else if (strcmp(format, "address") == 0)
```
Comment (olehermanse) [inline libpromises/evalfunction.c:583]: `StringEqual()`

File: libpromises/evalfunction.c [hunk 7]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+    {
+        return ADDRESS;
+    }
+    else if (strcmp(format, "key") == 0)
```
Comment (olehermanse) [inline libpromises/evalfunction.c:587]: `StringEqual()`

File: libpromises/evalfunction.c [hunk 8]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
 {
     Rlist *recent = NULL, *aged = NULL;
-    Item *ip;
     time_t now = time(NULL);
     double entrytime;
-    char address[CF_MAXVARSIZE]; // TODO: Could this be 1025 / NI_MAXHOST ?
+    char ret_host_data[CF_MAXVARSIZE]; // TODO: Could this be 1025 / NI_MAXHOST ?
 
-    for (ip = addresses; ip != NULL; ip = ip->next)
+    for (size_t i = 0; i < SeqLength(host_data); ++i)
```
Comment (olehermanse) [inline libpromises/evalfunction.c:604]:
```suggestion
    const size_t length = SeqLength(host_data);
    for (size_t i = 0; i < length; ++i)
```

File: libpromises/evalfunction.c [hunk 9]
```diff
@@ -629,34 +692,34 @@ static bool CallHostsSeenCallback(const char *hostkey, const char *address,
     char buf[PRINTSIZE(uintmax_t)];
     xsnprintf(buf, sizeof(buf), "%ju", (uintmax_t) quality->lastseen);
```
Comment (olehermanse) [inline libpromises/evalfunction.c:693]: Unnecessary to convert lastseen to string.

File: libpromises/evalfunction.c [hunk 10]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
     double entrytime;
-    char address[CF_MAXVARSIZE]; // TODO: Could this be 1025 / NI_MAXHOST ?
+    char ret_host_data[CF_MAXVARSIZE]; // TODO: Could this be 1025 / NI_MAXHOST ?
 
-    for (ip = addresses; ip != NULL; ip = ip->next)
+    for (size_t i = 0; i < SeqLength(host_data); ++i)
     {
-        if (sscanf(ip->classes, "%lf", &entrytime) != 1)
+        HostData *hd = SeqAt(host_data, i);
+        if (sscanf(hd->lastseen, "%lf", &entrytime) != 1)
```
Comment (olehermanse) [inline libpromises/evalfunction.c:607]: Unnecessary to convert lastssen back from string. (Just change the type in struct).

File: libpromises/evalfunction.c [hunk 11]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+    free(hd->lastseen);
+    free(hd);
+}
+
+typedef enum {
+    NAME,
+    ADDRESS,
+    KEY,
+    NONE
+} HostsSeenFormatOption;
```
Comment (olehermanse) [inline libpromises/evalfunction.c:575]:
```suggestion
} HostsSeenFieldOption;
```

https://docs.cfengine.com/docs/3.15/reference-functions-hostsseen.html

File: libpromises/evalfunction.c [hunk 12]
```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
             if (r)
             {
                 Log(LOG_LEVEL_DEBUG, "Purging from list of aged hosts.");
                 RlistDestroyEntry(&aged, r);
             }
 
             Log(LOG_LEVEL_DEBUG, "Adding to list of recent hosts.");
-            RlistPrependScalarIdemp(&recent, address);
+            RlistPrependScalarIdemp(&recent, ret_host_data);
         }
```
Comment (olehermanse) [inline libpromises/evalfunction.c:666]: As an optimization, you could do the filtering first, before rendering `ret_host_data`. But let's wait with this and do it in a follow-up PR once you have everything working.

File: libpromises/evalfunction.c [hunk 13]
```diff
@@ -585,41 +678,11 @@ static Rlist *GetHostsFromLastseenDB(Item *addresses, time_t horizon, bool retur
...
-        {
-            return FnReturnContext(false);
-        }
-    }
-
-    return FnReturnContext(true);
-}
-
-/*******************************************************************/
-
```
Comment (olehermanse) [inline libpromises/evalfunction.c:617]: Please move FnCallAnd back up to where it was, to make a nicer diff. Then make a separate commit where you only move it.

File: libpromises/evalfunction.c [hunk 14]
```diff
@@ -585,41 +678,11 @@ static Rlist *GetHostsFromLastseenDB(Item *addresses, time_t horizon, bool retur
...
 static bool CallHostsSeenCallback(const char *hostkey, const char *address,
                                   ARG_UNUSED bool incoming, const KeyHostSeen *quality,
                                   void *ctx)
 {
-    Item **addresses = ctx;
+    Seq *host_data = ctx;
```
Comment (olehermanse) [inline libpromises/evalfunction.c:685]:
```suggestion
    Seq *hosts = ctx;
```

File: libpromises/evalfunction.c [hunk 15]
```diff
@@ -629,34 +692,34 @@ static bool CallHostsSeenCallback(const char *hostkey, const char *address,
...
 
 static FnCallResult FnCallHostsSeen(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const Policy *policy, ARG_UNUSED const FnCall *fp, const Rlist *finalargs)
 {
-    Item *addresses = NULL;
+    Seq *host_data = SeqNew(1, HostDataFree);
 
     int horizon = IntFromString(RlistScalarValue(finalargs)) * 3600;
     char *hostseen_policy = RlistScalarValue(finalargs->next);
-    char *format = RlistScalarValue(finalargs->next->next);
+    char *format_str = RlistScalarValue(finalargs->next->next);
```
Comment (olehermanse) [inline libpromises/evalfunction.c:707]:
```suggestion
    char *field_str = RlistScalarValue(finalargs->next->next);
```

File: libpromises/evalfunction.c [hunk 16]
```diff
@@ -629,34 +692,34 @@ static bool CallHostsSeenCallback(const char *hostkey, const char *address,
...
+    HostsSeenFormatOption format = ValidateHostsSeenFormatOption(format_str);
```
Comment (olehermanse) [inline libpromises/evalfunction.c:708]:
```suggestion
    HostsSeenFieldOption format = ValidateHostsSeenFormatOption(field_str);
```

File: libpromises/evalfunction.c [hunk 17]
```diff
@@ -9206,7 +9269,7 @@ static const FnCallArg HOSTSSEEN_ARGS[] =
 {
     {CF_VALRANGE, CF_DATA_TYPE_INT, "Horizon since last seen in hours"},
     {"lastseen,notseen", CF_DATA_TYPE_OPTION, "Complements for selection policy"},
-    {"name,address", CF_DATA_TYPE_OPTION, "Type of return value desired"},
+    {"name,address,key", CF_DATA_TYPE_OPTION, "Type of return value desired"},
```
Comment (olehermanse) [inline libpromises/evalfunction.c:9272]: I think hostkey is more widely used.

```suggestion
    {"name,address,hostkey", CF_DATA_TYPE_OPTION, "Type of return value desired"},
```

**Review by olehermanse — CHANGES_REQUESTED** [review]
Good progress! First round of review comments, mostly around style and names :)

File: libpromises/evalfunction.c (hunk 3 above)
Comment (KotzaBoss) [reply libpromises/evalfunction.c:548]: refactor error, forgot to change it

File: libpromises/evalfunction.c (hunk 1 above)
Comment (KotzaBoss) [reply libpromises/evalfunction.c:538]: The reason is mostly cohesion. The GetHostsFromLastseenDB was above FnCallAnd and yet it wasnt used there. So i figured i would consolidate all the FnCallHostsSeen functions by moving that bellow And. I will revert it for now.

File: libpromises/evalfunction.c (hunk 5 above)
Comment (KotzaBoss) [reply libpromises/evalfunction.c:562]: Will follow the guidelines.
My reasoning was that (smth i wanted to write as a comment but forgot) that the HostData struct is like a consumable data type, therefore if one is created its expected to be destroyed as it was. Calling destroy means that hd and its fields are properly initialised.
This isnt some sophisticated idea its just some (possibly) idiomatic way of thinking i have but obviously not tested IRL.

File: libpromises/evalfunction.c (hunk 8 above)
Comment (KotzaBoss) [reply libpromises/evalfunction.c:604]: ouch, clumsy code there

File: libpromises/evalfunction.c (hunk 12 above)
Comment (KotzaBoss) [reply libpromises/evalfunction.c:666]: i was abit skeptical about this algorithm but true its not really relevant the current PR

File: libpromises/evalfunction.c (hunk 16 above)
Comment (KotzaBoss) [reply libpromises/evalfunction.c:708]: i should rename `format` -> `field` too?

File: libpromises/evalfunction.c (hunk 4 above)
Comment (KotzaBoss) [reply libpromises/evalfunction.c:545]: This links to the sprintf/sscanf comments bellow. I am not sure why this conversion has to happen too. But the [if check](https://github.com/cfengine/core/blob/master/libpromises/evalfunction.c#L522-L526) implies that something can go wrong but it looks like these sprintf/sscanf conversions are just asking for trouble.

Woudlnt it be ok to just have a `time_t` field and remove that if entirely?

File: libpromises/evalfunction.c (hunk 16 above)
Comment (olehermanse) [reply libpromises/evalfunction.c:708]: yep, edited.

File: libpromises/evalfunction.c [hunk 18]
```diff
@@ -509,67 +509,124 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
 }
 
-static Rlist *GetHostsFromLastseenDB(Item *addresses, time_t horizon, bool return_address, bool return_recent)
+typedef struct {
+    char *address;
+    char *hostkey;
+    time_t lastseen;
+} HostData;
+
+static HostData *HostDataNew(const char *address, const char *hostkey, const time_t lastseen)
```
Comment (vpodzime) [inline libpromises/evalfunction.c:518]: The `const` in `const time_t lastseen` is just confusing.

File: libpromises/evalfunction.c [hunk 19]
```diff
@@ -509,67 +509,124 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+{
+    HostData *data = xmalloc(sizeof(HostData));
+    data->address = xstrdup(address);
+    data->hostkey = xstrdup(hostkey);
```
Comment (vpodzime) [inline libpromises/evalfunction.c:522]: Please use `SafeStringDuplicate()` or add asserts for `address` and `hostkey` not being `NULL`.

File: libpromises/evalfunction.c [hunk 20]
```diff
@@ -509,67 +509,124 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+    data->lastseen = lastseen;
+    return data;
+}
+
+static void HostDataFree(HostData *hd)
+{
+    if(hd != NULL)
```
Comment (vpodzime) [inline libpromises/evalfunction.c:529]: Missing space.

File: libpromises/evalfunction.c [hunk 21]
```diff
@@ -619,44 +676,39 @@ static bool CallHostsSeenCallback(const char *hostkey, const char *address,
                                   ARG_UNUSED bool incoming, const KeyHostSeen *quality,
                                   void *ctx)
 {
-    Item **addresses = ctx;
-
     if (HostKeyAddressUnknown(hostkey))
     {
         return true;
     }
```
Comment (vpodzime) [inline libpromises/evalfunction.c:682]: Do we still want to omit these hosts in all the new cases?

File: libpromises/evalfunction.c [hunk 22]
```diff
@@ -509,67 +509,124 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
...
+}
+
+typedef enum {
+    NAME,
+    ADDRESS,
+    HOSTKEY,
+    NONE
+} HostsSeenFieldOption;
+
+static HostsSeenFieldOption ValidateHostsSeenFieldOption(const char *field)
```
Comment (vpodzime) [inline libpromises/evalfunction.c:544]: `Parse` would make more sense than `Validate` here. There's no validation happening.

File: libpromises/evalfunction.c [hunk 23]
```diff
@@ -619,44 +676,39 @@ static bool CallHostsSeenCallback(const char *hostkey, const char *address,
...
+    HostsSeenFieldOption field = ValidateHostsSeenFieldOption(field_str);
 
     Log(LOG_LEVEL_DEBUG, "Calling hostsseen(%d,%s,%s)",
-        horizon, hostseen_policy, format);
+        horizon, hostseen_policy, field_str);
 
-    if (!ScanLastSeenQuality(&CallHostsSeenCallback, &addresses))
+    if (!ScanLastSeenQuality(&CallHostsSeenCallback, host_data))
     {
         return FnFailure();
```
Comment (vpodzime) [inline libpromises/evalfunction.c:704]: Leaking `host_data` here.

File: libpromises/evalfunction.c [hunk 24]
```diff
@@ -619,44 +676,39 @@ static bool CallHostsSeenCallback(const char *hostkey, const char *address,
...
     }
 
-    Rlist *returnlist = GetHostsFromLastseenDB(addresses, horizon,
-                                               strcmp(format, "address") == 0,
+    Rlist *returnlist = GetHostsFromLastseenDB(host_data, horizon,
+                                               field,
                                                strcmp(hostseen_policy, "lastseen") == 0);
```
Comment (vpodzime) [inline libpromises/evalfunction.c:709]: Might as well change this to `StringEqual()`.

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise.

File: libpromises/evalfunction.c (hunk 21 above)
Comment (KotzaBoss) [reply libpromises/evalfunction.c:682]: You mean that we could ignore the fact that the address is uknown if we are asking for the shas? If so, we can easily move this check to GetHostsFromLastseenDB if the user asks for addersses.
```c
case ADDRESS:
{
    if (HostKeyAddressUnkown(hd->hostkey))
    {
        continue;
    }
    ...
}
```
Doesnt sound too unreasonable, however im not sure what the implications of an uknown address are...

Comment (vpodzime) [reply libpromises/evalfunction.c:682]: I think it makes sense to return the hostkey (sha) if its known and user asked for it.

Comment (KotzaBoss) [reply libpromises/evalfunction.c]: Should there be a "guard" in the [switch](https://github.com/cfengine/core/blob/0bfc9529b4cefa2a4c58d5e14229f9e8db2a901d/libpromises/evalfunction.c#L579-L595) if the case is NAME or ADDRESS to deal with any unknown addresses?

Comment (vpodzime) [reply libpromises/evalfunction.c]: As a user, I don't care about the implementation ;) But if I ask for hosts by hostkey, I'd expect to get all hosts that match the criteria and they hostkey is known.

**Comment by olehermanse** [comment]
Added onhold to this ticket as it will wait until after 3.18.0.

**Comment by KotzaBoss** [comment]
@cf-bottom, jenkins please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6837/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6837/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6837/

**Comment by vpodzime** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6852/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6852/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6852/

**Review by vpodzime — APPROVED** [review]
Looks good to me otherwise. Please make sure tests pass and documentation of the function gets updated with the new option

**Review by olehermanse — CHANGES_REQUESTED** [review]
The test is failing.

File: tests/acceptance/01_vars/02_functions/010.cf [hunk 25]
```diff
@@ -11,37 +11,45 @@ body common control
...
+      "ok_localhost_hostkey"
+        expression => reglist(@{init.hostkeys}, ${sys.key_digest});
+}
 
 bundle agent check
 {
-  vars:
-      "hosts" slist => { @{test.hosts} };
   classes:
-      "ok" expression => "any";        # XXX # I don't know how to test hostsseen!
```
Comment (nickanderson) [inline tests/acceptance/01_vars/02_functions/010.cf:37]: So, the prior test was not testing anything except perhaps that the agent didn't bail out on the function call. Here, it was just setting OK if any class is defined (which any is always defined).

`hostsseen()` pulls from the embedded databases where we have stored information about communication with peers ... the same information that `cf-key -s` pulls from.

Unless the host has actually communicated over the cfengine protocol there won't be anything in those data stores.
I don't think that this part of the test framework has any communication over the network with other cfengine hosts by default. There are some cf-serverd tests which do exercise the network. I think that you would need to force some network activity for this test to work.

It might fit better as a deployment test

Comment (vpodzime) [reply tests/acceptance/01_vars/02_functions/010.cf]: > It might fit better as a deployment test

Yup, I was thinking the same. Ideally we should add one DT testing various functions like this one -- https://tracker.mender.io/browse/CFE-3709

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6860/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6860/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6860/

**Comment by KotzaBoss** [comment]
@olehermanse I tweaked `cf_remote_demo.sh` script to install the proper [package](http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6861/PACKAGES_HUB_x86_64_linux_ubuntu_18/cfengine-nova-hub_3.19.0a.78cb63cf5~15291.ubuntu18_amd64.deb) on the hub:
```shell
# ... unaltered

cf-remote install --package http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6861/PACKAGES_HUB_x86_64_linux_ubuntu_18/cfengine-nova-hub_3.19.0a.78cb63cf5~15291.ubuntu18_amd64.deb --demo --bootstrap hub --hub hub

#cf-remote deploy --hub hub "$SCRIPTPATH/../../masterfiles"

# ... unaltered
```
 and wrote a simple test on the hub:
```
bundle agent main
{
  vars:
      "names" slist => hostsseen("inf", "lastseen", "name");
      "addrs" slist => hostsseen("inf", "lastseen", "address");
      "keys" slist => hostsseen("inf", "lastseen", "hostkey");

  classes:
      "HasHubName" expression => reglist(names, ${sys.fqhost});
      "HasHubIp" expression => reglist(addrs, ${sys.ipv4});
      "HasHubKey" expression => reglist(keys, ${sys.key_digest});

  reports:
      "name: ${names}";
      "addr: ${addrs}";
      "key: ${keys}";

    HasHubName::
      "Hub name '${sys.fqhost}' found";
    !HasHubName::
      "Hub name '${sys.fqhost}' NOT FOUND";

    HasHubIp::
      "Hub ip '${sys.ipv4}' found";
    !HasHubIp::
      "Hub ip '${sys.ipv4}' NOT FOUND";

    HasHubKey::
      "Hub key '${sys.key_digest}' found";
    !HasHubKey::
      "Hub key '${sys.key_digest}' NOT FOUND";
}
```
Output:
```
R: name: ip-172-31-30-234.eu-west-1.compute.internal
R: name: ip-172-31-23-29.eu-west-1.compute.internal
R: name: ip-172-31-30-176.eu-west-1.compute.internal
R: name: ip-172-31-25-224.eu-west-1.compute.internal
R: name: ip-172-31-25-145.eu-west-1.compute.internal
R: name: ip-172-31-19-53.eu-west-1.compute.internal
R: name: ip-172-31-16-115.eu-west-1.compute.internal
R: name: ip-172-31-24-73.eu-west-1.compute.internal
R: addr: 172.31.30.234
R: addr: 172.31.23.29
R: addr: 172.31.30.176
R: addr: 172.31.25.224
R: addr: 172.31.25.145
R: addr: 172.31.19.53
R: addr: 172.31.16.115
R: addr: 172.31.24.73
R: key: SHA=101056478647952e018ec32fc5a9a43cff3c8401a9bddcb332668862786bbc79
R: key: SHA=cd1adb305940fe3d8bbe65624b71e71301c2683025a5dabd517b16f0591d613d
R: key: SHA=784278f3bf7264494e1c22f826301c769b02f41852d1e17a73e9edafa53dcb85
R: key: SHA=8c45c160c155e207ad73ef6cbed7142bb8f8e660cba62212e79910854e5714a7
R: key: SHA=7bc884557c9cc24281ed92d74461f5b7cb0c335a175997808e537379ef913369
R: key: SHA=44beeea216ff3d263a31ccf18c5ac6a079bad4e1ebd68ab296ccf2cbeed4c71e
R: key: SHA=05315d4e22f231d3be68d03bff2ec13996dbc8a9dabf998726f6c86d012092e6
R: key: SHA=b16eb17e42b3a4d96c38fc6fbd2364f8f51208210c206fae5fd89f5a07733818
R: Hub name 'ip-172-31-30-234.eu-west-1.compute.internal' found
R: Hub ip '172.31.30.234' found
R: Hub key 'SHA=101056478647952e018ec32fc5a9a43cff3c8401a9bddcb332668862786bbc79' found
```
Is this close to being acceptable as a manual test?

**Comment by vpodzime** [comment]
> Is this close to being acceptable as a manual test?

Looks great to me. :+1:

**Review by vpodzime — APPROVED** [review]
ACK

**Review by olehermanse — APPROVED** [review]
